source/opt/loop_fusion.cpp

```diff
@@ -0,0 +1,731 @@
+      CollectChildren(child, loops);
+    }
+  }
+}
+
+// Return the set of locations accessed by |stores| and |loads|.
+std::set<ir::Instruction*> GetLocationsAccessed(
+    const std::map<ir::Instruction*, std::vector<ir::Instruction*>>& stores,
+    const std::map<ir::Instruction*, std::vector<ir::Instruction*>>& loads) {
+  std::set<ir::Instruction*> locations{};
+
+  for (const auto& kv : stores) {
+    locations.insert(std::get<0>(kv));
+  }
+
+  for (const auto& kv : loads) {
+    locations.insert(std::get<0>(kv));
+  }
+
+  return locations;
+}
+
+// Append all dependences from |sources| to |destinations| to |dependences|.
+void GetDependences(std::vector<DistanceVector>* dependences,
+                    LoopDependenceAnalysis* analysis,
+                    const std::vector<ir::Instruction*>& sources,
+                    const std::vector<ir::Instruction*>& destinations,
+                    size_t num_entries) {
+  for (auto source : sources) {
+    for (auto destination : destinations) {
+      DistanceVector dist(num_entries);
+      if (!analysis->GetDependence(source, destination, &dist)) {
+        dependences->push_back(dist);
+      }
+    }
+  }
+}
+
+// Apped all instructions in |block| to |instructions|.
+void AddInstructionsInBlock(std::vector<ir::Instruction*>* instructions,
+                            ir::BasicBlock* block) {
+  for (auto& inst : *block) {
+    instructions->push_back(&inst);
+  }
+
+  instructions->push_back(block->GetLabelInst());
+}
+
+}  // namespace
+
+bool LoopFusion::UsedInContinueOrConditionBlock(
+    ir::Instruction* phi_instruction, ir::Loop* loop) {
+  auto condition_block = loop->FindConditionBlock()->id();
+  auto latch_block = loop->GetLatchBlock()->id();
+  auto not_used = context_->get_def_use_mgr()->WhileEachUser(
+      phi_instruction,
+      [this, condition_block, latch_block](ir::Instruction* instruction) {
+        auto block_id = context_->get_instr_block(instruction)->id();
+        return block_id != condition_block && block_id != latch_block;
+      });
+
+  return !not_used;
+}
+
+void LoopFusion::RemoveIfNotUsedContinueOrConditionBlock(
+    std::vector<ir::Instruction*>* instructions, ir::Loop* loop) {
+  instructions->erase(
+      std::remove_if(std::begin(*instructions), std::end(*instructions),
+                     [this, loop](ir::Instruction* instruction) {
+                       return !UsedInContinueOrConditionBlock(instruction,
+                                                              loop);
+                     }),
+      std::end(*instructions));
+}
+
+bool LoopFusion::AreCompatible() {
+  // Check that the loops are in the same function.
+  if (loop_0_->GetHeaderBlock()->GetParent() !=
+      loop_1_->GetHeaderBlock()->GetParent()) {
+    return false;
+  }
+
+  // Make sure |loop_1_| has a preheader.
+  for (auto& loop : *context_->GetLoopDescriptor(containing_function_)) {
+    loop.GetOrCreatePreHeaderBlock();
```

Comment:
Scalar analysis can't handle loops without pre-headers, so I've left the creation here. I did update the fusion pass to correctly return `Status::SuccessWithChange` if the IR is modified because of this.

Reply:
Sure, but the issue I have is that you have a predicate that has side-effects here.  It's cleaner if the caller creates the preheader and then calls the predicate.  The predicate should return false if the loop has no pre-header.

Reply:
Do you mean the caller of `AreCompatible`?

Reply:
I've removed the side-effect in 52ec676.